cashweb-registry/src/store/topics.rs

```diff
@@ -0,0 +1,325 @@
+//! Contains `DbTopics`, allowing access to topic data.
+
+use crate::{
+    proto,
+    store::db::{Db, CF, CF_MESSAGES, CF_PAYLOADS},
+};
+use bitcoinsuite_core::{Hashed, Sha256};
+use bitcoinsuite_error::{ErrorMeta, Report, Result};
+use cashweb_payload::payload::{ParseSignedPayloadError, SignedPayload};
+use prost::Message;
+use rocksdb::{ColumnFamilyDescriptor, Direction, IteratorMode};
+use std::fmt::Debug;
+use thiserror::Error;
+
+type TopicPayload = SignedPayload<proto::BroadcastMessage>;
+
+/// Allows access to registry metadata.
+pub struct DbTopics<'a> {
+    db: &'a Db,
+    cf_messages: &'a CF,
+    cf_payloads: &'a CF,
+}
+
+/// Errors indicating some registry topic error.
+#[derive(Debug, Error, ErrorMeta)]
+pub enum DbTopicsError {
+    /// Foo
+    #[critical()]
+    #[error("bitcoinsuite error: {0}")]
+    PayloadDecode(#[from] ParseSignedPayloadError),
+
+    /// Foo
+    #[critical()]
+    #[error("bitcoinsuite error: {0}")]
+    ReportError(#[from] Report),
+
+    /// Foo
+    #[critical()]
+    #[error("RocksDB error: {0}")]
+    RocksDB(#[from] rocksdb::Error),
+
+    /// Foo
+    #[critical()]
+    #[error("Prost encode error: {0}")]
+    ProstEncode(#[from] prost::EncodeError),
+
+    /// Foo
+    #[critical()]
+    #[error("Prost decode error: {0}")]
+    ProstDecode(#[from] prost::DecodeError),
+
+    /// Foo
+    #[critical()]
+    #[error("Value not found in messages: {0}")]
+    MissingValue(String),
+
+    /// Foo
+    #[critical()]
+    #[error("Topic has too many separators: {0} > 10")]
+    TopicTooLong(usize),
+
+    /// Foo
+    #[critical()]
+    #[error("Topic contains invalid characters")]
+    TopicInvalidCharacters(),
+
+    /// Foo
+    #[critical()]
+    #[error("Topic contains empty segments")]
+    TopicInvalidSegments(),
+}
+
+/// Allows access to registry topics.
+impl<'a> DbTopics<'a> {
+    /// Create a new [`DbTopic`] instance.
+    pub fn new(db: &'a Db) -> Self {
+        let cf_messages = db.cf(CF_MESSAGES).unwrap();
+        let cf_payloads = db.cf(CF_PAYLOADS).unwrap();
+        DbTopics {
+            db,
+            cf_messages,
+            cf_payloads,
+        }
+    }
+
+    /// Put a serialized `Message` to database.
+    pub fn put_message(&self, timestamp: u64, message: &TopicPayload) -> Result<(), DbTopicsError> {
```

Comment:
Use `bitcoinsuite_error::Result` with ErrorMeta here, then you don't have to have such a crazy error type.

Reply:
ErrorMeta?

Reply:
Can you elaborate on what would need to happen. You helped me setup the error type above last time.